Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,640 @@
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    // Since these are normalized and to account for unsigned subtraction we have to do mantissa based comparison.
+
+    NSComparisonResult res = _compareMantissa(left, right);
+
+    if (res == NSOrderedSame) {
+        SET_TO_ZERO(result);
+        return NSCalculationNoError;
+    }
+
+    if (res == NSOrderedAscending) {
+        // R > L
+        NSCalculationError error = _subtraction(result, right, left);
+        if (error != NSCalculationNoError) {
+            INVALIDATE_DECIMAL(result);
+            return error;
+        }
+        result->_isNegative = right->_isNegative;
+        result->_exponent = right->_exponent;
+    } else {
+        NSCalculationError error = _subtraction(result, left, right);
+        if (error != NSCalculationNoError) {
+            INVALIDATE_DECIMAL(result);
+            return error;
+        }
+        result->_isNegative = left->_isNegative;
+        result->_exponent = left->_exponent;
+    }
+
+    NSDecimalCompact(result);
+    return NSCalculationNoError;
+}
+
+static NSCalculationError _subtraction(NSDecimal* result, const NSDecimal* left, const NSDecimal* right) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    unsigned int commonLength = std::min(left->_length, right->_length);
+
+    unsigned int overflow = 1;
+
+    // Since the decimal is normalized, we just subtract the mantissas
+    // subtract the numbers in the length
+    for (int index = 0; index < commonLength; index++) {
+        overflow += USHRT_MAX + ((left->_mantissa)[index] - (right->_mantissa)[index]);
+        (result->_mantissa)[index] = overflow;
+        overflow = overflow >> 0x10;
+    }
+
+    unsigned int length = right->_length;
+
+    // handle the rest of the length.
+    if (right->_length < left->_length) {
+        while (length < left->_length && overflow == 0) {
+            overflow = USHRT_MAX + (left->_mantissa)[length];
+            (result->_mantissa)[length] = overflow;
+            overflow = overflow >> 0x10;
+            length++;
+        }
+
+        memmove((result->_mantissa) + length, left->_mantissa + length, (left->_length - length) * sizeof(unsigned short));
+        length = (length < left->_length) ? left->_length : length;
+    }
+
+    result->_length = length;
+
+    return NSCalculationNoError;
+}
+
+static NSCalculationError _mantissaAddition(NSDecimalInternal* result, NSDecimalInternal* augend, unsigned short addend) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    // Add the addend to the mantissa
+    for (int index = 0; index < augend->decimal._length; ++index) {
+        unsigned int res = static_cast<unsigned int>((augend->_mantissa)[index]) + static_cast<unsigned int>(addend);
+        (result->_mantissa)[index] = res;
+        addend = res >> 0x10;
+    }
+
+    if (addend != 0) {
+        if (result->_extendedLen == augend->decimal._length) {
```

Comment:
confused - shouldn't this only be considered an overflow if we go past NSDecimalMaxSize (no more shorts to fill?)